client/lib/viewport/react-helpers.js

```diff
@@ -0,0 +1,120 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React, { useState, useEffect } from 'react';
+import { camelCase, upperFirst } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	isWithinBreakpoint,
+	addWithinBreakpointListener,
+	removeWithinBreakpointListener,
+	MOBILE_BREAKPOINT,
+	DESKTOP_BREAKPOINT,
+} from './index';
+
+/**
+ * React hook for getting the status for a breakpoint and keeping it updated.
+ *
+ * @param {String} breakpoint The breakpoint to consider.
+ *
+ * @returns {Boolean} The current status for the breakpoint.
+ */
+export function useBreakpoint( breakpoint ) {
+	const [ isActive, setIsActive ] = useState( isWithinBreakpoint( breakpoint ) );
+
+	function handleBreakpointChange( currentStatus ) {
+		setIsActive( currentStatus );
+	}
+
+	useEffect(() => {
+		const subscription = addWithinBreakpointListener( breakpoint, handleBreakpointChange );
+
+		return function cleanup() {
+			removeWithinBreakpointListener( breakpoint, subscription );
+		};
+	}, []);
```

Comment:
What if the `breakpoint` prop changes? A well-behaved hook should update the `isActive` value and also change the source of the subscription.

There is a ["subscribe to friend's online status" example](https://reactjs.org/docs/hooks-effect.html#explanation-why-effects-run-on-each-update) in React docs that implements the "resubscribe on friend prop change" behavior, but doesn't implement the `isOnline` state update when the observed friend changes.

That part is implemented in React's [`create-subscription`](https://github.com/facebook/react/blob/bb2939ccc23c9895d798f889d9c32848be43225e/packages/create-subscription/src/createSubscription.js#L70-L82) package, using `getDerivedStateFromProps`. But that's a render prop wrapper component, not a hook. I don't know what's the best way to adapt that technique to a `useSubscription( source )` hook.

Reply:
Yes, I suppose I can't force `breakpoint` to be a static thing; a component might want to dynamically change what breakpoint it's listening to depending on, say, user choice.

As far as I can tell, the only change that's needed here is to make the effect hook depend on `breakpoint`, right? That'll trigger cleanup and a new subscription. Or am I missing a weird edge-case?

Reply:
There indeed is a weird edge-case: if the `breakpoint` changes, the `isActive` value will have an old, incorrect value until the new listener actually fires for the first time.

It's not a blocker and we can fix as soon as we know how :smile: See also: https://github.com/facebook/react/issues/14988#issuecomment-468676819

Reply:
In that case, adding this to the top of `useEffect` might fix it?

```js
const currentStatus = isWithinBreakpoint( breakpoint );
  if ( isActive !== currentStatus ) {
    setIsActive( currentStatus );
  }
```

Reply:
Yes, that appears to be it. Added the change and a test for it!

Reply:
> In that case, modifying useEffect like so might fix it?

It's an acceptable solution for our case, but wouldn't be good enough for a general-purpose `create-subscription-hook` 🙂  Consider the following component:
```js
function ShowBreakpoint( { breakpoint } ) {
  const isActive = useBreakpoint( breakpoint );
  return `${ breakpoint }: ${ isActive }`;
}
```
For a screen width 1000px, `<ShowBreakpoint breakpoint='<480px' />` renders
```
<480px: false
```
Updating the `breakpoint` prop to `>960px` first renders
```
>960px: false
```
and that's not true! Only an instant later, when the effect is executed, calls `setState` and the component is quickly rerendered to:
```
>960px: true
```

Reply:
You're absolutely right! Thank you for pointing out yet another edge case.

I came up with a different implementation after a bunch of testing in a sandbox, and I think this one works correctly in every situation (famous last words, I know). Instead of doing the direct comparison and firing an update if it changes, it does things differently:

- The state is now a complex object containing both the breakpoint and the `isActive` status. We could keep it as two different states, but updating both would mean two renders, which I want to avoid in such a foundational hook.
- `handleBreakpointChange` sets all of the state in one go, and returns the previous state if nothing changes. Returning the previous state helps us avoid re-rendering if nothing changes, thanks to React's early bail.
- The hook now only returns the `isActive` status in state if `breakpoint` matches the one in state. If there's a mismatch, it means `breakpoint` was changed and the effect hasn't had a chance to run yet. In that case, it retrieves the current status from the viewport.